Reconnaissance demo tutorial

vmayoral opened the pull request:
Add preliminar files for the reconnaissance demo tutorial.

vmayoral committed:
First iteration on reconnaissance demo

- Add aztarna (from source for development)
- Add configuration files for demo
- Add README.md instructions

vmayoral committed:
reconnaissance demo update

- Update Dockerfile
- Complete README.md
- Update dir naming

vmayoral commented:
ping @thomas-moulard

vmayoral committed:
Fix WORKDIR back and minor typo correction

vmayoral commented:
Reviewed and fixed the issue with the working directory (thanks @ruffsl :)!). @thomas-moulard could you please review it again?

ruffsl commented on README.md:
```diff
@@ -53,6 +53,38 @@ You can first localize the robot by initializing the pose and then setting a nav
+  --gateway=10.0.0.1 \
+	--ip-range 10.0.0.192/27 \
+  --attachable \
+  overlay
```
Isn't `overlay` already a commonly/default network name used docker?
Might want to try if to name it something else, or see if the default `overlay` is enough.

vmayoral replied:
I originally wanted to name it ros2net (and avoid potential confusions) but I then realized that rocker is hardcoding the networking options making it a bit difficult to join other-than-predefined networks. Tried a few approaches (e.g. using https://github.com/sloretz/off-your-rocker#docker-run-args) but eventually decided to avoid submitting PRs to rocker (didn't wanted to risk not getting it soon enough or having part of the audience with previous versions).

`overlay` is indeed a networking driver type but no network named `overlay` by default shows up in my docker instalation (tested in OS X):

```bash
victor:~/turtlebot3_demo (master)$
| => docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
9ca6f66f5c4a        bridge              bridge              local
debdd4d51bf3        docker_gwbridge     bridge              local
d6b4095558bd        host                host                local
k7fwav1rlz1o        ingress             overlay             swarm
ad76f78c7ba0        none                null                local
```

Can you confirm this last bit please?

ruffsl replied:
Ahh, newer installs on ubuntu seem to have changes to only create a default bridge instead:

```
$ docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
16fa8b27fe17        bridge              bridge              local
52ba16a29500        host                host                local
e688be1ac788        none                null                local
```

> rocker is hardcoding the networking options

Hmm, that's annoying. 

https://github.com/osrf/rocker/blob/8f0da1f49c7240efc96ffa4ffa48101af580911e/src/rocker/cli.py#L37

Ok, let's stick with what you have working for now.

ruffsl replied:
In any case, I've opened a ticket about this here: https://github.com/osrf/rocker/issues/70

vmayoral replied:
Let me resolve this for now @ruffsl.

ruffsl commented on configs/reconnaissance_demo/common.conf:
```diff
@@ -0,0 +1,76 @@
+send-keys 'env | grep ROS' Enter
+
+#----------------------
+# create reconnaissance window
```
If you don't need run any new commands on startup, perhaps you could just add this window to the main common.conf one directory up instead of duplicating the entire config. I makes it a little harder to keep changes in sync.

vmayoral replied:
Noted. I didn't wanted to influence any other demos but knowing this is preferred, let me do it!

ruffsl replied:
Yea, I'd like to refactor AWS's PR to be included the common.config, or simple source it like so to add windows without altering the common config like so:

https://github.com/ros-swg/turtlebot3_demo/blob/424b79efb1baa217700d21b0ae6fdabff99abdad/configs/unsecure.conf#L9

You could do something like

``` bash
source $BYOBU_PREFIX/share/byobu/profiles/tmux

# setup reconnaissance session
new-session -s reconnaissance -n diagnostic -d
send-keys 'glances' Enter

# create reconnaissance window
new-window -n reconnaissance
send-keys 'aztarna -t ros2'

split-window -v
send-keys 'aztarna -t ros2 -d 0'

split-window -h
send-keys 'tshark -i eth0 -f "udp" -Y "rtps"'

# Select ROS2 reconnaissance pane
select-pane -t 0

source configs/common.conf
```

vmayoral replied:
All right, gave it a try @ruffsl. Removed the reconnaissance-specific config files and integrated the simple conf for reconnaissance into `common.conf`.

vmayoral committed:
Remove reconnaissance-specific config files, minor fixes

vmayoral committed:
Merge branch 'master' into master

ruffsl commented on README.md:
````diff
@@ -53,6 +53,38 @@ You can first localize the robot by initializing the pose and then setting a nav
+
+### Host/container insider attacker
+``` bash
+rocker rosswg/turtlebot3_demo "byobu -f configs/unsecure.conf attach"
````
```suggestion
rocker rosswg/turtlebot3_demo:roscon19 "byobu -f configs/unsecure.conf attach"
```

ruffsl replied:
Let's stay consistent and point the users to use the reserved `roscon19` tag from the docker repo to ensure they stick with the curated image for the tutorial.

ruffsl commented on README.md:
```diff
@@ -53,6 +53,38 @@ You can first localize the robot by initializing the pose and then setting a nav
+  overlay
+
+# in another Terminal (terminal 2), launch demo
+rocker --network overlay rosswg/turtlebot3_demo "byobu -f configs/unsecure.conf attach"
```
```suggestion
rocker --network overlay rosswg/turtlebot3_demo:roscon19 "byobu -f configs/unsecure.conf attach"
```

ruffsl commented on README.md:
```diff
@@ -53,6 +53,38 @@ You can first localize the robot by initializing the pose and then setting a nav
+rocker --network overlay rosswg/turtlebot3_demo "byobu -f configs/unsecure.conf attach"
+
+# in another Terminal (terminal 3), launch another container
+docker run -it --rm --network overlay --name aztarna rosswg/turtlebot3_demo /bin/bash
```
```suggestion
docker run -it --rm --network overlay --name aztarna rosswg/turtlebot3_demo:roscon19 /bin/bash
```

ruffsl commented on README.md:
```diff
@@ -53,6 +53,38 @@ You can first localize the robot by initializing the pose and then setting a nav
+rocker --network overlay rosswg/turtlebot3_demo "byobu -f configs/unsecure.conf attach"
+
+# in another Terminal (terminal 3), launch another container
+docker run -it --rm --network overlay --name aztarna rosswg/turtlebot3_demo /bin/bash
```
What does the user need to do/type after launching bash?

vmayoral replied:
I plan to walk audience through it but I've added a simple scan for completeness. 1dc426384a8c20270d885c01a4da495e3763c213.

ruffsl commented on README.md:
````diff
@@ -53,6 +53,38 @@ You can first localize the robot by initializing the pose and then setting a nav
+
+### Host/container insider attacker
+``` bash
+rocker rosswg/turtlebot3_demo "byobu -f configs/unsecure.conf attach"
````
Does this container need not be connected to the same overlay network?

vmayoral replied:
No, this one aims to demontrate the attack from inside of the robot (inside of the container).

ruffsl commented on README.md:
```diff
@@ -53,6 +53,38 @@ You can first localize the robot by initializing the pose and then setting a nav
 [![](media/localize.png)](media/localize.mp4)
 
+## Running the reconnaissance demos:
+
```
Might want to add some text describing what this section is about, and what it demonstrates, and what each step does. This would be helpful for folks just reading this later offline or after/beyond the workshop session.

vmayoral replied:
Good point, thanks! See 1dc426384a8c20270d885c01a4da495e3763c213

ruffsl commented on Dockerfile:
```diff
@@ -60,6 +60,25 @@ RUN . $TB3_OVERLAY_WS/install/setup.sh && \
     ros2 security generate_artifacts -k keystore \
       -p policies/tb3_gazebo_policy.xml
 
+# install tools for footprinting and pentesting
```
Sorry, just came to my mind, but could you move this chunk up before line 55, so that if we or folks need to regenerate permissions with a changed policy, they wouldn't have to break so much of the build cache that would otherwise apt install this again.

vmayoral replied:
Sure. Done at https://github.com/ros-swg/turtlebot3_demo/pull/19/commits/82b2cd6fc05c49f14d65a9d26839b0357706c6f8. I tested things afterwards and everything seems to be operating fine. A double check would be appreciated though. Ping @mikaelarguedas, @thomas-moulard.

mikaelarguedas replied:
I'm tuning out for the day due to poor internet, I can give it a shot tomorrow morning if it hasnt been validated by then

vmayoral committed:
Update README.md

Co-Authored-By: Ruffin <roxfoxpox@gmail.com>

vmayoral committed:
Update README.md

Co-Authored-By: Ruffin <roxfoxpox@gmail.com>

vmayoral committed:
Update README.md

Co-Authored-By: Ruffin <roxfoxpox@gmail.com>

vmayoral committed:
Optimize cache with changes,

Attend https://github.com/ros-swg/turtlebot3_demo/pull/19/files/fb054e1478e74ec9156ff360d68eb6df4ec558fa#r337690173

vmayoral committed:
Further improve reconnaissance section

Address comments and reviews. Credit to @ruffsl

vmayoral commented:
I reproduced everything and there're some issues with the secure demos. See https://github.com/ros-swg/turtlebot3_demo/issues/29

vmayoral committed:
Minor fix in Dockerfile for reconnaissance

ruffsl commented:
Perhaps you'd like to rebase from  https://github.com/ros-swg/turtlebot3_demo/pull/30 and test the rebuilt image again?

mikaelarguedas commented:
This branch conflicts with master preventing me from rebasing it on top of master for testing.
@vmayoral could you please update this branch with a rebased version of this PR ? thanks!

vmayoral commented:
> This branch conflicts with master preventing me from rebasing it on top of master for testing.
> @vmayoral could you please update this branch with a rebased version of this PR ? thanks!

Thanks for reviewing @mikaelarguedas. Interesting! I get the following which made me think that no conflicts applied

<img width="990" alt="imagen" src="https://user-images.githubusercontent.com/1375246/67558141-b6d0f700-f716-11e9-88c6-9ee48a755eda.png">

All right, let me rework this.
